pageserver/src/walreceiver.rs

```diff
@@ -218,14 +264,24 @@ fn walreceiver_main(
 
     let query = format!("START_REPLICATION PHYSICAL {}", startpoint);
 
-    let copy_stream = rclient.copy_both_simple(&query)?;
-    let mut physical_stream = ReplicationIter::new(copy_stream);
+    let copy_stream = runtime.block_on(replication_client.copy_both_simple(&query))?;
+    let physical_stream = ReplicationStream::new(copy_stream);
+    pin!(physical_stream);
 
     let mut waldecoder = WalStreamDecoder::new(startpoint);
 
     let mut walingest = WalIngest::new(&*timeline, startpoint)?;
 
-    while let Some(replication_message) = physical_stream.next()? {
+    while let Some(replication_message) = runtime.block_on(async {
+        tokio::select! {
+            replication_message = physical_stream.next() => {replication_message},
+            _ = &mut interrupt_receiver => {
+                info!("walreceiver interrupted");
+                None
+            }
+        }
+    }) {
```

Comment:
I wish we had a `CancellableReplicationStream: Iterator` which would hide all of this complexity, but it occurs to me that's too much to ask for.

Reply:
This feature looks generic enough and I've found a crate for that: https://github.com/jonhoo/stream-cancel Not sure that we should bring it for just one use case